### DISCUSSION TITLE
Synchronised scrolling

### DISCUSSION THREAD

**Discussion by fecet** [quarto-dev/quarto-cli#3648] [Feature Requests]
Can we have Synchronised scrolling like https://github.com/iamcco/markdown-preview.nvim in preview? That is extremely useful for me.

That should look like:
![screenshot](https://user-images.githubusercontent.com/5492542/47603494-28e90000-da1f-11e8-9079-30646e551e7a.gif)

**Comment by mcanouil** [comment]
Your request lacks context.
In which editor or IDE? (I would assume NeoVim)
Note that this is independent of Quarto CLI.

**Reply by fecet** [reply]
Of course quarto cli cannot do this on its own, it should be done with some editor. I can use almost any editor so if you guys are interested, you can develop in whatever way you like (say start with rstudio). But it would be even better if neovim could get some love.

**Comment by mcanouil** [comment]
May I suggest to re-post on <https://github.com/quarto-dev/quarto-nvim> (or for someone with more rights to transfer the discussion).
@jmbuhr is doing some work on the extension.

**Reply by jmbuhr** [reply]
This is an interesting proposition. It is indeed something where the cli (the `quarto preview` server in this case) and an editor have to communicate, so the discussion is in the right place, I believe. Right now I think there is no readily available method to tell the preview server to scroll to a place, though it does navigate e.g. to the last saved file in a project. However, I know that there are some requests you can send to the server that are not currently user-facing, so maybe there is also one with such capabilities. @cscheid would know more in this case.

After all, the reason that the markdown preview plugin shown by @fecet can do this is exactly because it is also the server.

**Reply by mcanouil** [reply]
Thanks for the clarification.
So for this to be more general, a "language server protocol" (LSP) could be implemented, but this would require more work on Quarto side.

**Reply by jmbuhr** [reply]
Yes and no. This is not what a language server will and can do, it is something the quarto preview server could do (the program that renders and up to date version of your qmd file and refreshes the browser).

A language server on the other hand provides things like hover documentation, code diagnostics, completion etc. and communicates via the language server protocol (LSP) that microsoft invented: https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#languageServerProtocol
Ultimately, it is a program running in the background to which your code editor can send requests and get responses, such as "my cursor is currently here in this document, can you give me text that I can show in a hover window with documentation for the symbol?" via https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_hover

Neovim and VS Code are both examples of editors that have a client to send requests to language servers built in.

**Reply by jimjam-slam** [reply]
It might also be useful to note that you can tell `quarto preview` to run an arbitrary command using the [`preview: serve` project option key](https://quarto.org/docs/reference/projects/options.html#serve) if you just want to bypass Quarto's built-in preview server and use something else (or potentially use Quarto's but also fire something else up in the background) 🥳

**Comment by cskeeters** [comment]
I'm spoiled by the *synchronized scrolling* capability in `markdown-preview.nvim` and `tinymist` (for typst).  It's a killer feature for text-based document development, IMHO.

If the preview server (quarto) is going to always be separate from the neovim plugin, the quarto has to develop an interface for instructing the browser to scroll that the neovim plugin could then utilize.  Has anyone looked at the feasibility of keeping track of the cursor through all of the transforms?  I don't know how the other tools do it, but I wish I did.

**Comment by luyiyun** [comment]
I have developed an nvim plug https://github.com/luyiyun/quarto-sync.nvim.git, which can realize the function of synchronous scrolling. Welcome to try it.
